core/src/test/java/org/bitcoinj/script/ScriptTest.java

```diff
@@ -494,4 +487,56 @@ public void getToAddress() {
     public void getToAddressNoPubKey() {
         ScriptBuilder.createP2PKOutputScript(ECKey.random()).getToAddress(BitcoinNetwork.TESTNET, false);
     }
+
+    List<List<String>> readScriptTestsJson(String resourcePath) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        JavaType type = mapper.getTypeFactory().constructCollectionType(List.class, mapper.getTypeFactory()
+                .constructCollectionType(List.class, String.class));
+        List<List<String>> list = mapper.readValue(getClass().getResourceAsStream(resourcePath), type);
+        return list.stream()
+                .filter(test -> test.size() > 1)    // Filter out comment entries
+                .collect(Collectors.toList());
+    }
+
+    List<TestEntry> readTransactionsJson(String resourcePath) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        JavaType type = mapper.getTypeFactory().constructCollectionType(List.class, JsonNode.class);
+        List<JsonNode> nodes = mapper.readValue(getClass().getResourceAsStream(resourcePath), type);
+        return  nodes.stream()
+                .filter(test -> !(test.isArray() && test.size() == 1 && test.get(0).isTextual()))
+                .map(n -> mapper.convertValue(n, TestEntry.class))
+                .collect(Collectors.toList());
+    }
+
+    static class TestEntry {
+        public final List<ScriptPubKeyEntry> scriptPubKeyEntries;
+        public final String transaction;
+        public final String verifyFlags;
+
+        @JsonCreator
+        public TestEntry(List<Object> entry) {
+            this((List<List<Object>>) entry.get(0), (String) entry.get(1), (String) entry.get(2));
+        }
+
+        TestEntry(List<List<Object>> scriptPubKeyEntries, String transaction, String verifyFlags) {
+            this.scriptPubKeyEntries = scriptPubKeyEntries.stream()
+                    .map(l -> new ScriptPubKeyEntry((String) l.get(0), (int) l.get(1), (String) l.get(2)))
+                    .collect(Collectors.toList());
+            this.transaction = transaction;
+            this.verifyFlags = verifyFlags;
+        }
+
+    }
+
+    static class ScriptPubKeyEntry {
+        public final String hash;
+        public final long index;
+        public final String script;
+
+        public ScriptPubKeyEntry(String hash, long index, String script) {
+            this.hash = hash;
+            this.index = index;
+            this.script = script;
+        }
+    }
```

Comment:
I think these should implement toString() so that the original JSON line is output when an error occurs (as before this change).

Reply:
This would already be implemented if we had `record` 🌈🦄

I will add it.

Reply:
Fixed and force-pushed!

Reply:
> This would already be implemented if we had `record` 🌈🦄

Actually, that's not quite true because the built-in `record` `toString()` would not use JSON format.